README.md
# phylocomr

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![R-check](https://github.com/ropensci/phylocomr/actions/workflows/R-check.yml/badge.svg)](https://github.com/ropensci/phylocomr/actions/workflows/R-check.yml)
[![cran
version](https://www.r-pkg.org/badges/version/phylocomr)](https://CRAN.R-project.org/package=phylocomr)
[![codecov](https://codecov.io/gh/ropensci/phylocomr/branch/master/graph/badge.svg)](https://app.codecov.io/gh/ropensci/phylocomr)
[![rstudio mirror
downloads](https://cranlogs.r-pkg.org/badges/phylocomr)](https://github.com/r-hub/cranlogs.app)

`phylocomr` gives you access to the [Phylocom C
library](https://github.com/phylocom/phylocom/), licensed under [BSD
2-clause](https://opensource.org/license/bsd-2-clause/)

## Package API

-   `ecovolve`/`ph_ecovolve` - interface to `ecovolve` executable, and a
    higher level interface
-   `phylomatic`/`ph_phylomatic` - interface to `phylomatic` executable,
    and a higher level interface
-   `phylocom` - interface to `phylocom` executable
-   `ph_aot` - higher level interface to `aot`
-   `ph_bladj` - higher level interface to `bladj`
-   `ph_comdist`/`ph_comdistnt` - higher level interface to comdist
-   `ph_comstruct` - higher level interface to comstruct
-   `ph_comtrait` - higher level interface to comtrait
-   `ph_pd` - higher level interface to Faith’s phylogenetic diversity

## A note about files

As a convenience you can pass ages, sample and trait data.frame’s, and
phylogenies as strings, to `phylocomr` functions. However, `phylocomr`
has to write these data.frame’s/strings to disk (your computer’s file
system) to be able to run the Phylocom code on them. Internally,
`phylocomr` is writing to a temporary file to run Phylocom code, and
then the file is removed.

In addition, you can pass in files instead of data.frame’s/strings.
These are not themselves used. Instead, we read and write those files to
temporary files. We do this for two reasons. First, Phylocom expects the
files its using to be in the same directory, so if we control the file
paths that becomes easier. Second, Phylocom is case sensitive, so we
simply standardize all taxon names by lower casing all of them. We do
this case manipulation on the temporary files so that your original data
files are not modified.

## Installation

Stable version:

    install.packages("phylocomr")

Development version:

    remotes::install_github("ropensci/phylocomr")

    library("phylocomr")
    library("ape")

## ecovolve

    ph_ecovolve(speciation = 0.05, extinction = 0.005, time_units = 50)

## phylomatic

    taxa_file <- system.file("examples/taxa", package = "phylocomr")
    phylo_file <- system.file("examples/phylo", package = "phylocomr")
    (taxa_str <- readLines(taxa_file))

    #> [1] "campanulaceae/lobelia/lobelia_conferta"          
    #> [2] "cyperaceae/mapania/mapania_africana"             
    #> [3] "amaryllidaceae/narcissus/narcissus_cuatrecasasii"

    (phylo_str <- readLines(phylo_file))

    #> [1] "(((((eliea_articulata,homalanthus_populneus)malpighiales,rosa_willmottiae),((macrocentrum_neblinae,qualea_clavata),hibiscus_pohlii)malvids),(((lobelia_conferta,((millotia_depauperata,(layia_chrysanthemoides,layia_pentachaeta)layia),senecio_flanaganii)asteraceae)asterales,schwenkia_americana),tapinanthus_buntingii)),(narcissus_cuatrecasasii,mapania_africana))poales_to_asterales;"

    ph_phylomatic(taxa = taxa_str, phylo = phylo_str)

    #> [1] "(lobelia_conferta:5.000000,(mapania_africana:1.000000,narcissus_cuatrecasasii:1.000000):1.000000)poales_to_asterales:1.000000;\n"
    #> attr(,"taxa_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/taxa_577357322292"
    #> attr(,"phylo_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/phylo_57731c7bcbf7"

use various different references trees

    library(brranching)
    library(ape)

    r2 <- ape::read.tree(text=brranching::phylomatic_trees[['R20120829']])
    smith2011 <- ape::read.tree(text=brranching::phylomatic_trees[['smith2011']])
    zanne2014 <- ape::read.tree(text=brranching::phylomatic_trees[['zanne2014']])

    # R20120829 tree
    taxa_str <- c(
      "asteraceae/bidens/bidens_alba",
      "asteraceae/cirsium/cirsium_arvense",
      "fabaceae/lupinus/lupinus_albus"
    )
    ph_phylomatic(taxa = taxa_str, phylo = r2)

    #> [1] "(((bidens_alba:13.000000,cirsium_arvense:13.000000):19.000000,lupinus_albus:27.000000):12.000000)euphyllophyte:1.000000;\n"
    #> attr(,"taxa_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/taxa_577350fa1f1c"
    #> attr(,"phylo_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/phylo_5773551090cc"

    # zanne2014 tree
    taxa_str <- c(
      "zamiaceae/dioon/dioon_edule",
      "zamiaceae/encephalartos/encephalartos_dyerianus",
      "piperaceae/piper/piper_arboricola"
    )
    ph_phylomatic(taxa = taxa_str, phylo = zanne2014)

    #> [1] "(((dioon_edule:121.744843,encephalartos_dyerianus:121.744850)zamiaceae:230.489838,piper_arboricola:352.234711)spermatophyta:88.058670):0.000000;\n"
    #> attr(,"taxa_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/taxa_577332926cb5"
    #> attr(,"phylo_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/phylo_57732b1ef903"

    # zanne2014 subtree
    zanne2014_subtr <- ape::extract.clade(zanne2014, node='Loganiaceae')
    zanne_subtree_file <- tempfile(fileext = ".txt")
    ape::write.tree(zanne2014_subtr, file = zanne_subtree_file)
    taxa_str <- c(
      "loganiaceae/neuburgia/neuburgia_corynocarpum",
      "loganiaceae/geniostoma/geniostoma_borbonicum",
      "loganiaceae/strychnos/strychnos_darienensis"
    )
    ph_phylomatic(taxa = taxa_str, phylo = zanne2014_subtr)

    #> [1] "((neuburgia_corynocarpum:32.807743,(geniostoma_borbonicum:32.036335,strychnos_darienensis:32.036335):0.771406):1.635496)loganiaceae:0.000000;\n"
    #> attr(,"taxa_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/taxa_57737ac12496"
    #> attr(,"phylo_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/phylo_57731e4932d0"

    ph_phylomatic(taxa = taxa_str, phylo = zanne_subtree_file)

    #> [1] "((neuburgia_corynocarpum:32.807743,(geniostoma_borbonicum:32.036335,strychnos_darienensis:32.036335):0.771406):1.635496)loganiaceae:0.000000;\n"
    #> attr(,"taxa_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/taxa_577357a70538"
    #> attr(,"phylo_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/phylo_57731647cc7d"

## aot

    traits_file <- system.file("examples/traits_aot", package = "phylocomr")
    phylo_file <- system.file("examples/phylo_aot", package = "phylocomr")
    traitsdf_file <- system.file("examples/traits_aot_df", package = "phylocomr")
    traits <- read.table(text = readLines(traitsdf_file), header = TRUE,
      stringsAsFactors = FALSE)
    phylo_str <- readLines(phylo_file)
    ph_aot(traits = traits, phylo = phylo_str)

    #> $trait_conservatism
    #> # A tibble: 124 × 28
    #>    trait trait.n…¹  node name    age ntaxa n.nodes tip.mn tmn.r…² tmn.r…³ tip.sd
    #>    <int> <chr>     <int> <chr> <dbl> <int>   <int>  <dbl>   <int>   <int>  <dbl>
    #>  1     1 traitA        0 a         5    32       2   1.75    1000    1000  0.440
    #>  2     1 traitA        1 b         4    16       2   1.75     647     660  0.447
    #>  3     1 traitA        2 c         3     8       2   1.75     700     688  0.463
    #>  4     1 traitA        3 d         2     4       2   1.5      264     959  0.577
    #>  5     1 traitA        4 e         1     2       2   1         65    1000  0    
    #>  6     1 traitA        7 f         1     2       2   2       1000     544  0    
    #>  7     1 traitA       10 g         2     4       2   2       1000     294  0    
    #>  8     1 traitA       11 h         1     2       2   2       1000     549  0    
    #>  9     1 traitA       14 i         1     2       2   2       1000     542  0    
    #> 10     1 traitA       17 j         3     8       2   1.75     648     716  0.463
    #> # … with 114 more rows, 17 more variables: tsd.ranklow <int>, tsd.rankhi <int>,
    #> #   node.mn <dbl>, nmn.ranklow <int>, nmn.rankhi <int>, nod.sd <dbl>,
    #> #   nsd.ranklow <int>, nsd.rankhi <int>, sstipsroot <dbl>, sstips <dbl>,
    #> #   percvaramongnodes <dbl>, percvaratnode <dbl>, contributionindex <dbl>,
    #> #   sstipvnoderoot <dbl>, sstipvnode <dbl>, ssamongnodes <dbl>,
    #> #   sswithinnodes <dbl>, and abbreviated variable names ¹​trait.name,
    #> #   ²​tmn.ranklow, ³​tmn.rankhi
    #> 
    #> $independent_contrasts
    #> # A tibble: 31 × 17
    #>     node name    age n.nodes contrast1 contrast2 contr…¹ contr…² contr…³ lowval1
    #>    <int> <chr> <dbl>   <int>     <dbl>     <dbl>   <dbl>   <dbl>   <dbl>   <dbl>
    #>  1     0 a         5       2     0         0       0       0.254    1.97    1.75
    #>  2     1 b         4       2     0         1.03    0       0.516    1.94    1.75
    #>  3     2 c         3       2     0.267     0.535   0       0        1.87    1.5 
    #>  4     3 d         2       2     0.577     0       1.15    0        1.73    1   
    #>  5     4 e         1       2     0         0       0.707   0        1.41    1   
    #>  6     7 f         1       2     0         0       0.707   0        1.41    2   
    #>  7    10 g         2       2     0         0       1.15    0        1.73    2   
    #>  8    11 h         1       2     0         0       0.707   0        1.41    2   
    #>  9    14 i         1       2     0         0       0.707   0        1.41    2   
    #> 10    17 j         3       2     0.267     0.535   0       0        1.87    1.5 
    #> # … with 21 more rows, 7 more variables: hival1 <dbl>, lowval2 <dbl>,
    #> #   hival2 <dbl>, lowval3 <dbl>, hival3 <dbl>, lowval4 <dbl>, hival4 <dbl>, and
    #> #   abbreviated variable names ¹​contrast3, ²​contrast4, ³​contrastsd
    #> 
    #> $phylogenetic_signal
    #> # A tibble: 4 × 5
    #>   trait  ntaxa varcontr varcn.ranklow varcn.rankhi
    #>   <chr>  <int>    <dbl>         <int>        <int>
    #> 1 traitA    32    0.054             1         1000
    #> 2 traitB    32    0.109             1         1000
    #> 3 traitC    32    0.622            51          950
    #> 4 traitD    32    0.011             1         1000
    #> 
    #> $ind_contrast_corr
    #> # A tibble: 3 × 6
    #>   xtrait ytrait ntaxa  picr  npos ncont
    #>   <chr>  <chr>  <int> <dbl> <dbl> <int>
    #> 1 traitA traitB    32 0.248  18.5    31
    #> 2 traitA traitC    32 0.485  27.5    31
    #> 3 traitA traitD    32 0      16.5    31

## bladj

    ages_file <- system.file("examples/ages", package = "phylocomr")
    phylo_file <- system.file("examples/phylo_bladj", package = "phylocomr")
    ages_df <- data.frame(
      a = c('malpighiales','salicaceae','fabaceae','rosales','oleaceae',
            'gentianales','apocynaceae','rubiaceae'),
      b = c(81,20,56,76,47,71,18,56)
    )
    phylo_str <- readLines(phylo_file)
    (res <- ph_bladj(ages = ages_df, phylo = phylo_str))

    #> [1] "((((((lomatium_concinnum:20.250000,campanula_vandesii:20.250000):20.250000,(((veronica_candidissima:10.125000,penstemon_paniculatus:10.125000)plantaginaceae:10.125000,justicia_oblonga:20.250000):10.125000,marsdenia_gilgiana:30.375000):10.125000):10.125000,epacris_alba-compacta:50.625000)ericales_to_asterales:10.125000,((daphne_anhuiensis:20.250000,syzygium_cumini:20.250000)malvids:20.250000,ditaxis_clariana:40.500000):20.250000):10.125000,thalictrum_setulosum:70.875000)eudicots:10.125000,((dendrocalamus_giganteus:27.000000,guzmania_densiflora:27.000000)poales:27.000000,warczewiczella_digitata:54.000000):27.000000)malpighiales:1.000000;\n"
    #> attr(,"ages_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/ages"
    #> attr(,"phylo_file")
    #> [1] "/var/folders/ss/2tpkp325521_kfgn59g44vd80000gn/T//RtmpeMkOMc/phylo_577376efe8a"

    plot(ape::read.tree(text = res))

![](man/figures/unnamed-chunk-9-1.png)

## Meta

-   Please [report any issues or
    bugs](https://github.com/ropensci/phylocomr/issues).
-   License: MIT
-   Get citation information for `phylocomr` in R doing
    `citation(package = 'phylocomr')`
-   Please note that this package is released with a [Contributor Code
    of Conduct](https://ropensci.org/code-of-conduct/). By contributing
    to this project, you agree to abide by its terms.


NEWS.md
phylocomr 0.3.4
===============

* Maintenance release to:
  - fix CRAN issues related to compiler warnings `[-Warray-parameter=]`, `[-Wstringop-overflow=]`, and `[-Wstringop-truncation]`

phylocomr 0.3.3
===============

* Maintenance release to:
  - fix CRAN issues related to compiler warnings
  - add new maintainer info (#33)


phylocomr 0.3.2
===============

### MINOR IMPROVEMENTS

* move readme image into man/figures (#30)

### BUG FIXES

* fix for gcc `-fno-common` (#29) (#31)

phylocomr 0.3.0
===============

### NEW FEATURES

* via (#26) (see below) - we now no longer use file paths passed in directly to functions, but instead write to temporary files to run with Phylocom so that we do not alter at all the users files. We note this in the README and package level manual file `?phylocomr`
* package gains new manual file `?phylocomr-inputs` that details the four types of inputs to functions and what format they are expected in, including how they differ for passing in data.frame's vs. file paths (#28)

### BUG FIXES

* for all data.frame traits inputs to fxns, check that the first column is called `name` (Phylocom doesn't accept anything else) (#27)
* fix was originally for `ph_aot()`, but realized this touches almost all functions: Phylocom is case sensitive. We were already making sure all taxon names in phylogenies (tips and nodes) were lowercased, and were lowercasing names in tables passed in, but were not fixing case in file paths passed in by the user. Now across all functions we make sure case is all lowercase for taxon names in any user inputs, so case problems should no longer be an issue. (#26) via @Jez-R

phylocomr 0.2.0
===============

### BUG FIXES

* two fixes for `ph_bladj()`: 1) now we lowercase the taxon name column in the ages data.frame before writing the data.frame to disk to avoid any mismatch due to case (we write the phylogeny to disk with lowercased names); 2) bladj expects the root node name from the phyologeny to be in the ages file; we now check that (#25)


phylocomr 0.1.4
===============

### BUG FIXES

* fix examples (#22)
* improve class checks in internal code, swap `inherits` for `class` (#23)
* small fix to use of fread in C lib; check that fread worked, and if not if it was an EOF error or other error (#24)


phylocomr 0.1.2
===============

### MINOR IMPROVEMENTS

* fixes for failed checks on Solaris

### BUG FIXES

* fix to internals of all functions that handle a phylogeny. `ph_phylomatic` was working fine with very simple trees in all lowercase. we now internally lowercase all node and tip labels, on any phylogeny inputs (phylo object, newick string, file path (read, then re-write back to disk)). phylomatic wasn't working with any uppercase labels.


phylocomr 0.1.0
===============

### NEW FEATURES

* released to CRAN


R/bladj.R
#' bladj
#'
#' Bladj take a phylogeny and fixes the root node at a specified age,
#' and fixes other nodes you might have age estimates for. It then sets all
#' other branch lengths by placing the nodes evenly between dated nodes,
#' and between dated nodes and terminals (beginning with the longest
#' 'chains').
#'
#' @export
#' @param ages (data.frame/character) ages data.frame, or path to an ages
#' file. required. column names do not matter, and are discarded anyway.
#' the first column must be the node names, and the second column the node
#' ages. See Details.
#' @template phylo
#' @return newick string with attributes for where ages and phylo files
#' used are stored
#' @details See [phylocomr-inputs] for expected input formats
#' @section Common Errors:
#' A few issues to be aware of:
#' 
#' - the ages table must have a row for the root node with an age estimate.
#' bladj will not work without that. We attempt to check this but can only
#' check it if you pass in a phylo object; there's no easy way to parse a
#' newick string without requiring ape
#' - bladj is case sensitive. internally we lowercase all tip and node labels
#' and taxon names in your ages file to avoid any case sensitivity problems
#' 
#' @examples \dontrun{
#' ages_file <- system.file("examples/ages", package = "phylocomr")
#' phylo_file <- system.file("examples/phylo_bladj", package = "phylocomr")
#'
#' # from data.frame
#' ages_df <- data.frame(
#'   a = c('malpighiales','eudicots','ericales_to_asterales','plantaginaceae',
#'         'malvids', 'poales'),
#'   b = c(81, 20, 56, 76, 47, 71)
#' )
#' phylo_str <- readLines(phylo_file)
#' (res <- ph_bladj(ages = ages_df, phylo = phylo_str))
#' if (requireNamespace("ape")) {
#'   library(ape)
#'   plot(read.tree(text = res))
#' }
#'
#' # from files
#' ages_file2 <- file.path(tempdir(), "ages")
#' write.table(ages_df, file = ages_file2, row.names = FALSE,
#'   col.names = FALSE, quote = FALSE)
#' phylo_file2 <- tempfile()
#' cat(phylo_str, file = phylo_file2, sep = '\n')
#' (res <- ph_bladj(ages_file2, phylo_file2))
#' if (requireNamespace("ape")) {
#'   library(ape)
#'   plot(read.tree(text = res))
#' }
#'
#' # using a ape phylo phylogeny object
#' x <- read.tree(text = phylo_str)
#' if (requireNamespace("ape")) {
#'   library(ape)
#'   plot(x)
#' }
#' 
#' (res <- ph_bladj(ages_file2, x))
#' if (requireNamespace("ape")) {
#'   library(ape)
#'   tree <- read.tree(text = res)
#'   plot(tree)
#' }
#' }

ph_bladj <- function(ages, phylo) {
  assert(ages, c("data.frame", "character"))
  assert(phylo, c("phylo", "character"))

  ages <- ages_check(ages)
  check_root_node(phylo, ages)
  phylo <- phylo_check(phylo)

  cdir <- getwd()
  bdir <- dirname(ages)
  stopifnot(bdir == dirname(phylo))
  setwd(bdir)
  on.exit(setwd(cdir))

  out <- suppressWarnings(
    phylocom(c(
      "bladj",
      "-f", basename(phylo)
    ), intern = TRUE)
  )[1]
  phylocom_error(out)
  attr(out, "ages_file") <- ages
  attr(out, "phylo_file") <- phylo
  return(out)
}

check_root_node <- function(x, y) {
  if (!inherits(x, "phylo")) return()
  x$node.label <- tolower(x$node.label)
  ages_nodes <- utils::read.table(y, stringsAsFactors = FALSE)[,1]
  phylo_root <- x$node.label[1]
  if (nzchar(phylo_root)) {
    if (!phylo_root %in% ages_nodes) {
      warning(
        sprintf("bladj may fail; the root node '%s' is not in the ages nodes",
          phylo_root), immediate. = TRUE)
    }
  }
}
